### PR TITLE
Extract Requirement Fix

### PR DESCRIPTION
**Pull request by r1ft4469**
You were writing the the extracts twice the first line of code was being overwritten by the second.

### CODE REVIEW

File: package.js [hunk 1]
```diff
@@ -694,44 +678,40 @@ class PathToTarkovController {
...
                 }
             }
         }
-
-        const database = DatabaseServer.tables;
-
-        Object.keys(this.config.exfiltrations).forEach(mapName => {
-            const extractPoints = Object.keys(this.config.exfiltrations[mapName]);
-            database.locations[mapName].base.exits = extractPoints.map(createExitPoint(this.entrypoints[mapName]));
-        });
```
Comment (guillaumearm) [inline package.js:734]: Do not remove this

Comment (r1ft4469) [reply package.js:734]: I have with no ill reactions to the server or spawning bots or loading into maps?

Comment (guillaumearm) [reply package.js:734]: The problem is here, just keep this lines behind a `if (this.config.remove_all_exfils_restrictions)` should do the job

Comment (guillaumearm) [reply package.js:734]: > I have with no ill reactions to the server or spawning bots or loading into maps?

Not sure to understand your message, but you just removed a feature right here.

Comment (r1ft4469) [reply package.js]: if your talking about the all extracts to be open the code already does that.

File: package.js [hunk 2]
```diff
@@ -392,7 +392,6 @@ const createSpawnPoint = (pos, rot, entrypoints) => {
         "BotZoneName": ""
     }
 }
-
```
Comment (guillaumearm) [inline package.js:395]: add empty line here

File: package.js [hunk 3]
```diff
@@ -408,7 +407,6 @@ const createExitPoint = (entrypoints) => (name) => {
         "ExfiltrationTime": 10
     }
 }
-
```
Comment (guillaumearm) [inline package.js:411]: add empty line here

File: package.js [hunk 4]
```diff
@@ -694,44 +692,49 @@ class PathToTarkovController {
         for (let i in locations) {
             if (isValidMap(i)) {
                 for (let x in locations[i].base.exits) {
-                    // Remove extracts restrictions
```
Comment (guillaumearm) [inline package.js:697]: put back the comment

File: package.js [hunk 5]
```diff
@@ -694,44 +692,49 @@ class PathToTarkovController {
...
-                        }
-                        if (locations[i].base.exits[x].RequirementTip !== '') {
-                            locations[i].base.exits[x].RequirementTip = '';
-                        }
-                        if (locations[i].base.exits[x].RequiredSlot) {
-                            delete locations[i].base.exits[x].RequiredSlot;
-                        }
-                    }
-
-                    // Make all extractions available to extract
```
Comment (guillaumearm) [inline package.js:719]: put back this comment too

**Review by guillaumearm — CHANGES_REQUESTED** [review]
Just few minor changes before we can merge, thanks again for contrib ;-)

**Comment by r1ft4469** [comment]
Updated the Code. I understand you are adding entrypoints onto the map with this but you do not need them unless you are using custom then you will have to pass multiple augments and not just the name.

**Comment by guillaumearm** [comment]
> but you do not need them unless you are using custom then you will have to pass multiple augments and not just the name.

I don't get it, entrypoints are fixed by map but the exit points are customs (from the config file)

Please note this:
```js
extractPoints.map(createExitPoint(this.entrypoints[mapName]))
```
is equivalent to this
```js
extractPoints.map(ep => createExitPoint(this.entrypoints[mapName])(ep))
```

because `createExitPoint` is a [curried function](https://javascript.info/currying-partials)

**Comment by r1ft4469** [comment]
everything should be updated now for how you want.

**Review by r1ft4469 — COMMENTED** [review]
Added Back

**Review by r1ft4469 — COMMENTED** [review]
Added

File: package.js [hunk 6]
```diff
@@ -694,44 +694,59 @@ class PathToTarkovController {
...
+                            if (locations[i].base.exits[x].RequiredSlot) {
+                               delete locations[i].base.exits[x].RequiredSlot;
+                            }
+                        }
                     }
                 }
             }
         }
+        
+        if (!this.config.remove_all_exfils_restrictions) {
```
Comment (guillaumearm) [inline package.js:742]: This line break the default config requirements.

When I ran the game with the default config file, exfils are corrects.

Here, we want to set our customs exits points when this option is true.

Sorry the code is not very comprehensible here and don't reflect what `remove_all_exfils_restrictions` really means.

The correct line of code is:
```js
if (this.config.remove_all_exfils_restrictions) {
```

**Review by guillaumearm — CHANGES_REQUESTED** [review]
Last issue and we're good to go ^^

File: package.js (hunk 6 above)
Comment (r1ft4469) [reply package.js:742]: sounds good I was running test now so we just need to flip that check?

**Comment by guillaumearm** [comment]
Oh actually I know what you means here.

the `remove_all_exfils_restrictions` name is a bit weird and the second part override the rest.

the correct name name would be `override_exfils` or something like that.

Actually the part that loops on exits and remove exfils restrictions is dead code and should be removed.

It's not the job of this mod to do this, use AllExtractionsAvailable mod if you want to remove exfil condition

**Comment by r1ft4469** [comment]
right the code base will still do it with the first changes but we could just remove it and set the the chance to 100 at start.

File: package.js (hunk 6 above)
Comment (guillaumearm) [reply package.js:742]: Exactly ! see [my comment here](https://github.com/guillaumearm/PathToTarkov/pull/2#issuecomment-1063213817) for more details

File: package.js [hunk 7]
```diff
@@ -694,44 +694,59 @@ class PathToTarkovController {
...
+                            
+                            if (locations[i].base.exits[x].Id !== '') {
+                                locations[i].base.exits[x].Id = '';
+                            }
+                            
+                            if (locations[i].base.exits[x].Count !== 0) {
+                                locations[i].base.exits[x].Count = 0;
+                            }
+                            
+                            if (locations[i].base.exits[x].PassageRequirement !== "None") {
```
Comment (guillaumearm) [inline package.js:721]: Actually all this part should be removed, [see this comment for more infos](https://github.com/guillaumearm/PathToTarkov/pull/2#issuecomment-1063213817)

File: package.js [hunk 8]
```diff
@@ -694,44 +694,59 @@ class PathToTarkovController {
...
-                        }
-                        if (locations[i].base.exits[x].RequiredSlot) {
-                            delete locations[i].base.exits[x].RequiredSlot;
-                        }
-                    }
-
-                    // Make all extractions available to extract
-                    if (locations[i].base.exits[x].Name !== "EXFIL_Train") {
+                    if (locations[i].base.exits[x].Name !== "EXFIL_Train" && !locations[i].base.exits[x].Name.includes("lab") || locations[i].base.exits[x].Name === "lab_Vent") {
+                        // Make all extractions available to extract
```
Comment (guillaumearm) [inline package.js:698]: Actually all this part should be removed, [see this comment for more infos](https://github.com/guillaumearm/PathToTarkov/pull/2#issuecomment-1063213817)

File: package.js [hunk 9]
```diff
@@ -694,44 +694,59 @@ class PathToTarkovController {
...
                         if (locations[i].base.exits[x].Chance !== 100) {
                             locations[i].base.exits[x].Chance = 100;
```
Comment (guillaumearm) [inline package.js:700]: this part should be removed too, [see this comment for more infos](https://github.com/guillaumearm/PathToTarkov/pull/2#issuecomment-1063213817)

**Review by guillaumearm — CHANGES_REQUESTED** [review]
I think we are good after that.

File: package.js (hunk 7 above)
Comment (r1ft4469) [reply package.js:721]: sounds good then suggest AllExtractionsAvailable mod to remove exfil restrictions

**Comment by guillaumearm** [comment]
> right the code base will still do it with the first changes but we could just remove it and set the the chance to 100 at start or we can leave it it and let it set the restrictions like I have it.

Hum 🤔 and what about renaming `remove_all_exfils_restrictions` to a `bypass_exfils_override` (false by default)

So it's more clean, on your side you can just bypass the exfils override done by PathToTarkov and add mod you want to tweak exfil on your own, what do you think ?

**Comment by guillaumearm** [comment]
Additional note: Regarding of how [`createExitPoint`](https://github.com/guillaumearm/PathToTarkov/blob/master/package.js#L396) is implemented, there is no restrictions at all when PTT override the exfils

**Comment by r1ft4469** [comment]
I like that that would leave some functionality out but you would have a cleaner code base.

**Comment by guillaumearm** [comment]
> I like that that would leave some functionality out but you would have a cleaner code base.

That's not the only point, it's important to have a way to configure PTT to NOT TOUCH exfils at all in order to avoid incompatibilities with other mods.

File: package.js [hunk 10]
```diff
@@ -694,44 +694,22 @@ class PathToTarkovController {
...
-                        }
-                    }
-
                     // Make all extractions available to extract
-                    if (locations[i].base.exits[x].Name !== "EXFIL_Train") {
-                        if (locations[i].base.exits[x].Chance !== 100) {
-                            locations[i].base.exits[x].Chance = 100;
-                        }
+                    if (locations[i].base.exits[x].Chance !== 100) {
+                        locations[i].base.exits[x].Chance = 100;
```
Comment (guillaumearm) [inline package.js:699]: Remove this.

Actually you can remove the whole loop (and the isValidMap function if not used anymore)

**Comment by r1ft4469** [comment]
> > I like that that would leave some functionality out but you would have a cleaner code base.
> 
> That's not the only point, it's important to have a way to configure PTT to NOT TOUCH exfils at all in order to avoid incompatibilities with other mods.

Very true. Do we even need to touch chance then or just let CreateExitPoint do it?

File: package.js [hunk 11]
```diff
@@ -694,44 +694,22 @@ class PathToTarkovController {
...
                     }
                 }
             }
         }
+        
+        if (this.config.remove_all_exfils_restrictions) {
```
Comment (guillaumearm) [inline package.js:705]: Change this by:

```js
if (!this.config.bypass_exfils_override ) {
```

This time, there is a bang ;-)

I'll add this field in the mod documentation.

**Review by guillaumearm — CHANGES_REQUESTED** [review]
Hope this is my last request changes 😂 

Anyway, sorry for the mess , and thanks again

**Comment by r1ft4469** [comment]
Should all be good now. Your code base is great btw.

**Review by guillaumearm — APPROVED** [review]
Neat, I'll make 1 test or 2 and I release soon.

**Comment by guillaumearm** [comment]
> Should all be good now. Your code base is great btw.

Yeah but there is no unit tests, no tooling (like eslint) and it would be nice to split the code because the package.js begin to grow :-)

**Comment by r1ft4469** [comment]
The downsides of working inside of mod engines.

**Comment by guillaumearm** [comment]
FYI 2.7.0 has been released on the hub

**Pull request merged**